prow/plugins/reward-owners/reward-owners.go

```diff
@@ -0,0 +1,201 @@
+	removedUserRe = regexp.MustCompile(`(?m)^-\s+-(.+?)$`)
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(PluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(_ *plugins.Configuration, _ []config.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: fmt.Sprintf("The reward-owners plugin watches in %s and %s files for modifications and welcomes new approvers and reviewers.", ownersconfig.DefaultOwnersFile, ownersconfig.DefaultOwnersAliasesFile),
+	}
+	return pluginHelp, nil
+}
+
+type githubClient interface {
+	IsCollaborator(owner, repo, login string) (bool, error)
+	IsMember(org, user string) (bool, error)
+	AddLabel(org, repo string, number int, label string) error
+	CreateComment(owner, repo string, number int, comment string) error
+	CreateReview(org, repo string, number int, r github.DraftReview) error
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+	GetPullRequest(org, repo string, number int) (*github.PullRequest, error)
+	RemoveLabel(owner, repo string, number int, label string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	BotUserChecker() (func(candidate string) bool, error)
+}
+
+type info struct {
+	org          string
+	repo         string
+	repoFullName string
+	number       int
+}
+
+func handlePullRequest(pc plugins.Agent, pre github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened && pre.Action != github.PullRequestActionReopened && pre.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+
+	var skipTrustedUserCheck bool
+	for _, r := range pc.PluginConfig.Owners.SkipCollaborators {
+		if r == pre.Repo.FullName {
+			skipTrustedUserCheck = true
+			break
+		}
+	}
+
+	prInfo := info{
+		org:          pre.Repo.Owner.Login,
+		repo:         pre.Repo.Name,
+		repoFullName: pre.Repo.FullName,
+		number:       pre.Number,
+	}
+	return handle(pc.GitHubClient, pc.GitClient, pc.Logger, &pre.PullRequest, prInfo, pc.PluginConfig.TriggerFor(pre.Repo.Owner.Login, pre.Repo.Name), skipTrustedUserCheck, pc.PluginConfig.OwnersFilenames)
+}
+
+func handle(ghc githubClient, gc git.ClientFactory, log *logrus.Entry, pr *github.PullRequest, info info, triggerConfig plugins.Trigger, skipTrustedUserCheck bool, resolver ownersconfig.Resolver) error {
+	org := info.org
+	repo := info.repo
+	number := info.number
+	filenames := resolver(org, repo)
+
+	// Get changes.
+	changes, err := ghc.GetPullRequestChanges(org, repo, number)
+	if err != nil {
+		return fmt.Errorf("error getting PR changes: %v", err)
+	}
+
+	// List added and removed users from OWNERS and OWNERS_ALIASES files.
+	addedUsers := sets.NewString()
+	removedUsers := sets.NewString()
+	for _, change := range changes {
+		if (filepath.Base(change.Filename) == filenames.Owners || filepath.Base(change.Filename) == filenames.OwnersAliases) && change.Status != github.PullRequestFileRemoved {
```

Comment:
IMO loading the YAML and doing a structured difference instead of using regex to parse a patch of a structured dataset might be nicer. Do we want to post this in all cases, or only if this is someone's *first* time becoming an OWNER? If so, we may just clone the new commit and traverse the owners set using our libraries to do the diff and figure out if this is a new member or someone owning new directories.

Reply:
first time becoming an OWNER would be best but didn't want to make the scope too complicated out of the gate 🙇🏻